### DISCUSSION TITLE
What is the best way to make the document id more accessible?

### DISCUSSION THREAD

**Discussion by KCErb** [tylim88/FirelordJS#98] [Q&A] [answered]
I've started doing this with my types and I'm not sure that it is the best.

```js
type GroupWithoutID = {
  labels: { [uid: string]: string },
  participants: string[],
};
export type GroupMetaType = MetaTypeCreator<
  GroupWithoutID,
  'groupsCollection'
>
export type Group = GroupWithoutID & ID;
```

The reason for this is that when I get data from the database I have to do something like this:

```js
group = document.data()
```

but if I want the document id of the group I need to call

```js
document.id
```

usually in my code, when I want data, I just want a flat object:

```js
{...document.data(), id: document.id}
```

so that I don't have to carry the document and the data around separately. The result is that I end up needing three types as you see at the beginning of this post.

1.  A type without ID on it to create the metatype
2. The id type
3. a type with id on it for normal use.

Is this an OK pattern? Am I missing anything?

**Comment by tylim88** [comment, marked as answer]
a more accurate and reusable type would be
```ts
type GroupWithoutID = {
  labels: { [uid: string]: string },
  participants: string[],
};
export type GroupMetaType = MetaTypeCreator<
  GroupWithoutID,
  'groupsCollection'
>
export type Group = GroupMetaType['read'] & { id: GroupMetaType['docID'] };
```

`data()` always return `read` type

normally document id type is string but it is possible to assign literal type like `abc${string}123` to document id type, so it is better to read document id type from meta type 

Meta type is your single source of truth, more about meta type https://firelordjs.com/guides/metatype#index-accessor

**Reply by KCErb** [reply]
Awesome, this is exactly what I was looking for (and assumed it existed). FirelordJS is so great 🤩
